nft_ingester/src/rocks_db.rs

```diff
@@ -81,3 +81,70 @@ pub async fn restore_rocksdb(config: &IngesterConfig) -> Result<(), BackupServic
     info!("restore_rocksdb fin");
     Ok(())
 }
+
+pub enum RocksDbManager {
+    Primary(Arc<Storage>),
+    Secondary(RocksDbSecondaryDuplicateMode),
+}
+
+impl RocksDbManager {
+    /// Instantiate a new RocksDbManager with primary storage
+    pub fn new_primary(primary: Arc<Storage>) -> Self {
+        RocksDbManager::Primary(primary)
+    }
+
+    /// Instantiate a new RocksDbManager with secondary storage
+    pub fn new_secondary(primary: Storage, secondary: Storage) -> Self {
+        RocksDbManager::Secondary(RocksDbSecondaryDuplicateMode {
+            rocks_db_instance: [primary.into(), secondary.into()],
+            current_rocks_db: AtomicUsize::new(0),
+        })
+    }
+
+    /// Returns the current storage instance
+    ///
+    /// It's always the same storage for primary mode
+    /// For secondary mode, it will return a storage instance
+    /// based on it's sync status with the primary storage
+    pub fn acquire(&self) -> Arc<Storage> {
+        match self {
+            RocksDbManager::Primary(storage) => storage.clone(),
+            RocksDbManager::Secondary(duplicate_mode) => duplicate_mode.rocks_db_instance
+                [duplicate_mode.current_rocks_db.load(Ordering::Relaxed)]
+            .clone(),
+        }
+    }
+
+    /// Syncronize secondary rocksdb with primary
+    ///
+    /// One of the DB will be blocked while the other one is processing request
+    pub async fn catch_up(&self) {
+        const SLEEP_TIME_MS: u64 = 10;
+
+        match self {
+            RocksDbManager::Primary(_) => {}
+            RocksDbManager::Secondary(duplicate_mode) => {
+                let free_node_idx =
+                    (duplicate_mode.current_rocks_db.load(Ordering::Relaxed) + 1) % 2;
+                let free_node = &duplicate_mode.rocks_db_instance[free_node_idx];
+
+                while Arc::<rocks_db::Storage>::strong_count(&free_node) > 1 {
```

Comment:
maybe also pass shutdown_rx here? we do not need to catch up DB when stop signal is sent to app

Reply:
`catch_up` will be fired only if clone_rx (main:229) is empty. Isn't it sufficient?

Reply:
yeap, the comment is about a situation you described